modules/core/js/src/main/scala/net/message/Scram.scala

```diff
@@ -0,0 +1,136 @@
+
+  private implicit class StringOps(val value: String) extends AnyVal {
+    def bytesUtf8: ByteVector = ByteVector.view(value.getBytes(java.nio.charset.StandardCharsets.UTF_8))
+  }
+
+  private val normalize = js.Dynamic.global.require("saslprep").asInstanceOf[String => String]
+
+  def clientFirstBareWithRandomNonce: ByteVector = {
+    val nonce = {
+      val arr = new Array[Byte](32)
+      java.security.SecureRandom.getInstanceStrong().nextBytes(arr)
+      ByteVector.view(arr).toBase64
+    }
+    clientFirstBareWithNonce(nonce)
+  }
+
+  def clientFirstBareWithNonce(nonce: String): ByteVector =
+    s"n=,r=${nonce}".bytesUtf8
+
+  case class ServerFirst(nonce: String, salt: ByteVector, iterations: Int)
+  object ServerFirst {
+    private val Pattern = """r=([\x21-\x2B\x2D-\x7E]+),s=([A-Za-z0-9+/]+={0,2}),i=(\d+)""".r
+
+    def decode(bytes: ByteVector): Option[ServerFirst] =
+      utf8.decodeValue(bytes.bits).toOption.flatMap {
+        case Pattern(r, s, i) =>
+          Some(ServerFirst(r, ByteVector.fromValidBase64(s), i.toInt))
+        case _ => 
+          None
+      }
+  }
+
+  case class ClientProof(value: String)
+
+  case class ClientFinalWithoutProof(channelBinding: String, nonce: String) {
+    override def toString: String = s"c=$channelBinding,r=$nonce"
+    def encode: ByteVector = toString.bytesUtf8
+    def encodeWithProof(proof: ClientProof): ByteVector = (toString ++ s",p=${proof.value}").bytesUtf8
+  }
+
+  case class Verifier(value: ByteVector)
+
+  case class ServerFinal(verifier: Verifier)
+  object ServerFinal {
+    private val Pattern = """v=([A-Za-z0-9+/]+={0,2})""".r
+    def decode(bytes: ByteVector): Option[ServerFinal] =
+      utf8.decodeValue(bytes.bits).toOption.flatMap {
+        case Pattern(v) =>
+          Some(ServerFinal(Verifier(ByteVector.fromValidBase64(v))))
+        case _ => 
+          None
+      }
+  }
+
+  private val crypto = js.Dynamic.global.require("crypto")
+
+  private def bufferToByteVector(buffer: js.Dynamic): ByteVector =
+    ByteVector.view(TypedArrayBuffer.wrap(
+      buffer.asInstanceOf[js.typedarray.ArrayBuffer],
+      buffer.byteOffset.asInstanceOf[Int],
+      buffer.byteLength.asInstanceOf[Int]
+    ))
+
+  private def HMAC(key: ByteVector, str: ByteVector): ByteVector = {
+    val mac = crypto.createHmac("sha256", key.toByteBuffer.arrayBuffer())
+    mac.update(str.toByteBuffer.arrayBuffer())
+    bufferToByteVector(mac.digest())
+  }
+
+  private def H(input: ByteVector): ByteVector = input.digest("SHA-256")
+
+  private def Hi(str: String, salt: ByteVector, iterations: Int): ByteVector = {
+    // TODO It is unfortunate that we have to use a sync API here when an async is available
+    // To make the change here will require running an F[_]: Async up the hiearchy    
+    val salted = crypto.pbkdf2Sync(str, salt.toByteBuffer.arrayBuffer(), iterations, 8 * 32, "sha256")
+    bufferToByteVector(salted)
+  }
```

Comment:
FYI.

Reply:
Can you elaborate? PBKDF-2 should be completely CPU bound

Reply:
Node.js offers an async [version of this method with a callback](https://nodejs.org/api/crypto.html#crypto_crypto_pbkdf2_password_salt_iterations_keylen_digest_callback). I assume they have a good reason for doing so.

My best guess would be that you are indeed right that it is CPU-bound, but it's CPU intensive enough that it would detrimentally block a single-threaded Node.js app.